05-Mining-Protocol.md

```diff
@@ -1,101 +1,171 @@
+In other words, any rules imposed on transaction selection/ordering by miners not described in the rest of this document may result in invalid work.
 
-This separation vastly simplifies the protocol implementation for clients that don’t support extended channels, as they only need to implement the subset of protocol messages related to standard channels (see Mining Protocol Messages for details).
+Mining servers MUST assign a unique subset of the search space to each mining device, otherwise the mining devices will waste energy with overlapping search.
 
-### 5.1.1 Standard Channels
+This protocol explicitly expects that upstream server software is able to manage the size of the hashing space correctly for its clients and can provide new and unique Jobs quickly enough, based on the hashpower of each client.
 
-Standard channels are intended to be used by end mining devices.
+The protocol defines two main types of Jobs: **Standard Jobs** and **Extended Jobs**.
 
-The size of the search space for one standard channel (header-only mining) for one particular value in the `nTime` field is `2^(NONCE_BITS + VERSION_ROLLING_BITS) = ~280Th`, where `NONCE_BITS = 32` and `VERSION_ROLLING_BITS = 16`.
-This is a guaranteed space before `nTime` rolling (or changing the Merkle root).
+This separation vastly simplifies the protocol implementation for clients that don’t support Extended Jobs, as they only need to implement the subset of protocol messages related to Standard Jobs (see Mining Protocol Messages for details).
 
-The protocol dedicates all directly modifiable bits (`version`, `nonce`, and `nTime`) from the block header to one mining channel.
-This is the smallest assignable unit of search space by the protocol.
-The client which opened the particular channel owns the whole assigned space and can split it further if necessary (e.g. for multiple hashing boards and for individual chips etc.).
+Additionally, a Job (either Standard or Extended) also could be potentially labeled as a **Future Job**.
 
-### 5.1.2 Extended Channels
+### 5.1.1 Standard Jobs
 
-Extended channels are intended to be used by proxies.
-Upstream servers which accept connections and provide work MUST support extended channels.
-Clients, on the other hand, do not have to support extended channels, as they MAY be implemented more simply with only standard channels at the end-device level.
-Thus, upstream servers providing work MUST also support standard channels.
+Standard Jobs are restricted to fixed Merkle Roots. We call this header-only mining (HOM).
 
-The size of search space for an extended channel is `2^(NONCE_BITS+VERSION_ROLLING_BITS+extranonce_size*8)` per `nTime` value.
+The size of the search space for one Standard Job, given a fixed `nTime` field, is `2^(NONCE_BITS + BIP320_VERSION_ROLLING_BITS) = ~280Th`, where `NONCE_BITS = 32` and `BIP320_VERSION_ROLLING_BITS = 16`.
```

Comment:
I like more the phrase:
"The size of the search space for one Standard Job is `2^(NONCE_BITS + BIP320_VERSION_ROLLING_BITS) = ~280Th`, per `nTime` value, where `NONCE_BITS = 32` and `BIP320_VERSION_ROLLING_BITS = 16`."

Reply:
According to this, does it mean we should think about the fact that newer machines are already surpassing this 280Th/s limit? (look at [S21 hydro](https://shop.bitmain.com/product/detail?pid=00020240718151523125gE0aB5Es065D) or [S21 XP hydro](https://shop.bitmain.com/product/detail?pid=000202406200044065805TcPB8mt0682))

Reply:
HOM is not intended for powerful machine.

Reply:
hmm so [this](https://github.com/plebhash/sv2-spec/blob/3e986cd4e6447e0ff9fa486ad762c66a8ecfb771/05-Mining-Protocol.md?plain=1#L39) phrase is wrong?

> All SV2 mining devices are restricted to Standard Jobs. This is a big difference from legacy SV1 mining devices, where rolling extranonces is a common feature.

Reply:
yes

Reply:
Can some powerful SV2 Mining Device open Extended Channels and receive `NewExtendedMiningJob`?

Reply:
this was a bit confusing to me, because it makes me think: "what is the purpose of HOM Standard Jobs after all?"

and based on this question https://github.com/stratum-mining/sv2-spec/pull/98#discussion_r1746809495, I believe @GitGab19 also shares this feeling

I thought about @Fi3 answer and here's my current perspective on the trade-offs of Standard Jobs:

---

rolling Extranonces is an "advanced feature" that makes firmware more complex and harder to maintain, so from a business perspective, it should be avoidable if possible.

so a way to correct the phrase above would be:
> All SV2 mining devices with hashrate under 280 TH/s should be restricted to HOM via Standard Jobs. This is a big difference from legacy SV1 mining devices, where rolling extranonces is a common feature. Hashing space optimization should be achieved upstream.
>
> If the SV2 mining device hashrate is above 280 TH/s, then it should support non-HOM via Extended Jobs.

---

hopefully the spec authors can confirm this in the near future (when we ask them to review this PR)

Reply:
I think HOM can be used for almost any machine, provided it doesn't exceed the hash rate by orders of magnitude.

The device is supposed to receive new mining jobs periodically. Every new job resets the ntime value again to the `min_ntime` for a given job. So unless the time doesn't roll hours into the future within the usual 30 second window, I don't think it's a problem.

Problem with rolling nTime too much ahead of the true time is equivalent to a problem of "not receiving a new mining job frequently enough".
It's difficult to draw some hard line. It's probably better to be a bit vague about that. Say something like "be sane about nTime rolling or otherwise your work will be rejected".

quick calculation:
Assume we allow nTime to roll 10 minutes into the future for a given mining job that is being received every 30 seconds. In order to get to that point we would need 5.6 PH/s (=280 * 600 / 30). I don't think we are going to see such devices in foreseeable future (simple device controlled by a single controller - more complex devices split across multiple machines need to open an extended channel).

Reply:
thanks for the clarification!

I'll tag @rrybarczyk here since this topic was part of a discussion we had not long ago, while we went through this PR.

When we had this discussion, my understanding was that HOM over Standard Jobs were designed to cater for Mining Devices with hashpower below 280 TH/s. This clarifies that this is not the case at all. HOM is perfectly suitable for almost any kind of Mining Device, as long as jobs are updated frequently.

But now that leaves me wondering: is there any scenario where Mining Devices would open Extended Channels with an upstream?